zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/SpotRestorationTool.java

```diff
@@ -0,0 +1,276 @@
+    }
+    this.dataDir = dataDir;
+    this.zk = zk;
+    this.snapLog = new FileTxnSnapLog(dataDir, dataDir);
+    this.targetZNodePath = targetZNodePath;
+    this.restoreRecursively = restoreRecursively;
+    this.messages = Lists.newArrayList();
+  }
+
+  /**
+   * Run the spot restoration.
+   * 1. If a node exists in restored data tree but not in the zk server, create the node in the server
+   * 2. If a node exists in zk server but not the restored data tree, show messages ask user to manually delete the node
+   * 3. If a node exists in both zk server and restored data tree, show message ask user to take care of the node
+   * @throws IOException
+   */
+  public void run() throws IOException {
+    DataTree dataTree = new DataTree();
+    long zxidRestored =
+        snapLog.restore(dataTree, new ConcurrentHashMap<>(), new FileTxnSnapLog.PlayBackListener() {
+          @Override
+          public void onTxnLoaded(TxnHeader hdr, Record rec, TxnDigest digest) {
+            // Do nothing since we are trying to build a data tree in memory, not in actual zk server
+          }
+        });
+
+    DataNode originalNode = dataTree.getNode(targetZNodePath);
+    if (originalNode == null) {
+      return;
+    }
+    Stat originalNodeStat = new Stat();
+    originalNode.copyStat(originalNodeStat);
+
+    LOG.info(
+        "The restored zk data tree is constructed. The highest zxid restored is: " + zxidRestored);
+    LOG.info("Requested path node data: " + Arrays.toString(originalNode.getData()));
+    LOG.info(
+        "Requested path node children: " + Arrays.toString(originalNode.getChildren().toArray()));
+    LOG.info("Requested path node stat: " + originalNodeStat.toString());
+    String requestMsg = "Do you want to restore to this point? Enter \"yes\" or \"no\".";
+    String yesMsg = "Restoring ZooKeeper server...";
+    String noMsg = "Restoration aborted. Nothing has been changed.";
+    if (getUserConfirmation(requestMsg, yesMsg, noMsg)) {
+      traverseNode(zk, dataTree, targetZNodePath, true);
+      printExitMessages();
+    }
+  }
+
+  /**
+   * 1. Restore the node value in zk server;
+   * 2. check if there are deprecated nodes (exist in zk server but not in restored data tree) in its child nodes;
+   * 3. Do the above operations for all of its child nodes in the restored data tree
+   * @param zk A client connection to zk server
+   * @param dataTree The restored zk data tree
+   * @param path The znode path to restore
+   * @param targetNode True if the node being restored is the target znode that user specified;
+   *                   false if it's a child node of the target node
+   */
+  private void traverseNode(ZooKeeper zk, DataTree dataTree, String path, boolean targetNode) {
+    if (!restoreNode(zk, dataTree, path, targetNode) || !restoreRecursively) {
+      // This node is skipped, there's no need to traverse its child nodes
+      return;
+    }
```

Comment:
This might work, but to reduce confusion and make the intention clear, I prefer something like the following:

```
singleZNodeRestore(zk, dataTree, ...);
if (!restoreRecursively) {
  // Non-recursive spot restoration, so no need to proceed further
  return;
}
```

What do you think?